bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -209,7 +212,8 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{
 			RequeueAfter: DefaultTokenTTL / 2,
 		}, nil
-	case config.Status.DataSecretName != nil:
+	// Return if we've already generated the config. No need to regenerate it.
+	case config.Status.Ready:
```

Comment:
Minor improvement,

you could collapse the two above for better readbility:

```go
	// Status is ready means a config has been generated.
	// The BootstrapToken has been generated for a join.
	case config.Status.Ready:
		// When the infrastructure is NOT ready indicates the token in the config has not been consumed and it may need a refresh.
		if config.Spec.JoinConfiguration != nil && config.Spec.JoinConfiguration.Discovery.BootstrapToken != nil && !configOwner.IsInfrastructureReady() {
			token := config.Spec.JoinConfiguration.Discovery.BootstrapToken.Token

			remoteClient, err := r.remoteClient(r.Client, cluster, r.scheme)
			if err != nil {
				log.Error(err, "error creating remote cluster client")
				return ctrl.Result{}, err
			}

			log.Info("Refreshing token until the infrastructure provider has a chance to consume it")
			if err := refreshToken(remoteClient, token); err != nil {
				// It would be nice to re-create the bootstrap token if the error was "not found", but we have no way to update the Machine's bootstrap data
				return ctrl.Result{}, errors.Wrapf(err, "failed to refresh bootstrap token")
			}
			// NB: this may not be sufficient to keep the token live if we don't see it before it expires, but when we generate a config we will set the status to "ready" which should generate an update event
			return ctrl.Result{
				RequeueAfter: DefaultTokenTTL / 2,
			}, nil
		}

		return ctrl.Result{}, nil
```

Reply:
ooooo

yes! I like this